src/resources/uart_esp32.cc

```diff
@@ -17,95 +17,584 @@
-    UartResource* uart_res = static_cast<UartResource*>(r);
-    SystemEventSource::instance()->run([&]() -> void {
-      FATAL_IF_NOT_ESP_OK(uart_driver_delete(uart_res->port()));
-    });
+  void on_unregister_resource(Resource* r) override {
+    auto uart_res = reinterpret_cast<UartResource*>(r);
     uart_ports.put(uart_res->port());
   }
 
-  uint32_t on_event(Resource* r, word data, uint32_t state) {
-    switch (data) {
-      case UART_DATA:
-        state |= kReadState;
-        break;
+  uint32_t on_event(Resource* r, word data, uint32_t state) override;
+};
+
+UART_ISR_INLINE void RxTxBuffer::return_buffer(uint8_t* buffer) {
+  // No blocking calls ever on TxBuffer, so ignore last parameter
+  vRingbufferReturnItemFromISR(ring_buffer(), buffer, null);
+}
+
+void TxBuffer::write(const uint8_t* buffer, uint16_t length, uint8_t break_length) {
+  {
+    SpinLocker locker(&spinlock_);
 
-      case UART_BREAK:
-        // Ignore.
-        break;
+    TxTransferHeader header = {
+        .break_length_ = static_cast<uint8_t>(break_length),
+        .remaining_data_length_ = static_cast<uint16_t>(length)
+    };
 
-      default:
-        state |= kErrorState;
-        break;
+    if (xRingbufferSend(ring_buffer(), &header, sizeof(TxTransferHeader), 0) == pdFALSE) {
+      abort();
     }
 
-    return state;
+    if (xRingbufferSend(ring_buffer(), buffer, length, 0) == pdFALSE) {
+      abort();
+    }
   }
-};
 
-MODULE_IMPLEMENTATION(uart, MODULE_UART);
+  uart()->enable_interrupt_index(UART_TOIT_INTR_TXFIFO_EMPTY);
+}
+
+UART_ISR_INLINE uint8_t TxBuffer::read_break() {
+  IsrSpinLocker locker(&spinlock_);
+
+  uint8_t break_length = transfer_header_.break_length_;
+  transfer_header_.break_length_ = 0; // To avoid it being read again and the break continuing...
+  return break_length;
+}
+
+UART_ISR_INLINE uint8_t* TxBuffer::read(size_t *received, size_t max_length) {
+  IsrSpinLocker locker(&spinlock_);
+
+  if (transfer_header_.remaining_data_length_ == 0 && transfer_header_.break_length_ == 0) {
+    // No more data in the latest package. Need to read header
+    size_t header_received;
+    size_t read = 0;
+    do { // As the header is multiple bytes, it could be split over the edge of the ring buffer
+      void* header_data = xRingbufferReceiveUpToFromISR(ring_buffer(), &header_received, sizeof(TxTransferHeader) - read);
+      if (!header_data) {
+        // TxBuffer is empty
+        return null;
+      }
+      memcpy(reinterpret_cast<uint8_t*>(&transfer_header_) + read, header_data, header_received);
+      // No blocking calls ever on TxBuffer, so ignore last parameter
+      vRingbufferReturnItemFromISR(ring_buffer(), header_data, null);
+      read += header_received;
+    } while (read < sizeof(TxTransferHeader));
+  }
+
+  size_t maximum_to_receive = max_length;
+  if (transfer_header_.remaining_data_length_ < maximum_to_receive) maximum_to_receive = transfer_header_.remaining_data_length_;
```

Comment:
```suggestion
  size_t maximum_to_receive = Utils::min(max_length, transfer_header_.remaining_data_length_);
```

Reply:
I "fixed" it, however Utils::min requires the arguments to be identical types, so I had to do: 
```
uword maximum_to_receive = Utils::min(max_length, static_cast<uword>(transfer_header_.remaining_data_length_));
```